### ISSUE TITLE
[DX] Allow consistent HTML element selection in UIs

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1818]
Both Views and Blocks allow you to rewrite their markup, but they allow different (and sometimes inadequate) tags. Let's make these consistent and complete the list of allowed options.

In views, the list of allowed elements is stored in config `views.settings.json`.  

```
'views_field_rewrite_elements' => array(
      '' => t('- Use default -'),
      '0' => t('- None -'),
      'div' => 'DIV',
      'span' => 'SPAN',
      'h1' => 'H1',
      'h2' => 'H2',
      'h3' => 'H3',
      'h4' => 'H4',
      'h5' => 'H5',
      'h6' => 'H6',
      'p' => 'P',
      'strong' => 'STRONG',
      'em' => 'EM',
    ),
```

We could move this variable out of views` and use it in all locations: views, layouts, and blocks.

**Comment by BWPanda** [comment]
If so, I recommend removing the first two options and leaving them up to the calling function to add as necessary.

**Comment by klonos** [comment]
...less places to be translating the first 2 options too 👍 

So, a function that returns an array? Something like:

```php
function system_html_tags() {
  $html_tags = array(
    '' => t('- Use default -'),
    '0' => t('- None -'),
    'div' => 'DIV',
    'span' => 'SPAN',
    'h1' => 'H1',
    'h2' => 'H2',
    'h3' => 'H3',
    'h4' => 'H4',
    'h5' => 'H5',
    'h6' => 'H6',
    'p' => 'P',
    'strong' => 'STRONG',
    'em' => 'EM',
  );

  return $html_tags;
}
```

...and then do:

```php
views_field_rewrite_elements' => system_html_tags(),
```

Not sure if it'd be best to exclude options, or instead be unsetting array elements as required whenever we call this. Like:

```php
$html_tags = system_html_tags();
unset(...);
unset(...);

...
  'views_field_rewrite_elements' => $html_tags,
...
```

**Comment by BWPanda** [comment]
See how this goes... https://github.com/backdrop/backdrop/pull/3503

**Comment by klonos** [comment]
I'm calling this a DX clean-up task rather than a new feature. Recommending it gets in on a bug fix release.

**Comment by jenlampton** [comment]
> I'm calling this a DX clean-up task rather than a new feature.

uh...  I'm not sure that will fly. But we'll see :)

Because this was a setting before, I don't think we should hard-code the list of allowed tags without at least allowing it to be altered.  *reads PR*  Of course, @BWPanda already thought of that 👍 

What do we do about the scenario where people had changed the views setting `field_rewrite_elements` before this PR goes in? They're going to see a regression until they have the ability to add the alter hook.

What if we moved the existing views variable to system config, rather than just deleting it. If it exists, we use that instead? For new installs, it won't exist.

**Comment by BWPanda** [comment]
>  reads PR Of course, @BWPanda already thought of that 👍

Actually that was @docwilmot's suggestion, so kudos to him for that!

**Comment by BWPanda** [comment]
> What do we do about the scenario where people had changed the views setting field_rewrite_elements before this PR goes in?

Good point. I've updated the PR with one solution (slightly different from @jenlampton's suggestion. See what you think...

**Comment by jenlampton** [comment]
I love this new solution!  

But I still think `$form['field_rewrite_elements']` needs to be removed from the views settings form. :)

**Comment by BWPanda** [comment]
> But I still think $form['field_rewrite_elements'] needs to be removed from the views settings form. :)

Hmm, missed that! Do you know why it was there in the first place? All I can tell is that the config value was loaded into the form, then saved again when the form was submitted. But can't see it actually being used anywhere...

Anyway, PR updated to remove it altogether. Also fixed a bug in my last commit.

**Comment by klonos** [comment]
> Anyway, PR updated to remove it altogether.

Yeah, if we won't be storing it on config for new sites, there's absolutely no need for that to be there.

The code looks good 👍 but I think we're missing the html elements select for layout regions, which should also use the new function or the hook:

![image](https://user-images.githubusercontent.com/2423362/105328188-75a8ea80-5bd8-11eb-9bab-a8b608feb953.png)

**Comment by klonos** [comment]
...same with felxible template rows I think:

![image](https://user-images.githubusercontent.com/2423362/105328561-d59f9100-5bd8-11eb-862f-fe64b0784112.png)

**Comment by BWPanda** [comment]
@klonos Hmm, those places use a very different set of elements... This issue was originally about replacing *duplicate* lists of elements. I don't mind using the new function in other parts of core, but I'm wondering how that'd work then.

Will all places get a list of all elements (e.g. Views' wrapper tag for field label shows options for `aside`, `header` & `footer`)? Or do we somehow give each part of core access to only part of the list?

**Comment by klonos** [comment]
🤔

**Comment by jenlampton** [comment]
> Do you know why it was there in the first place?

My guess is that was an easy way to get the setting saved into variables?

> Will all places get a list of all elements (e.g. Views' wrapper tag for field label shows options for aside, header & footer)? Or do we somehow give each part of core access to only part of the list?

We have an argument for `inline`, what if we changed that to a more generic `types` array, and the values could be `inline`, `block`, `layout`, and `all`. This would allow us to specify `layout` for layout and get  `header`, `asside`, and `footer` when needed.

**Comment by klonos** [comment]
> We have an argument for `inline`, what if we changed that to a more generic `types` array...

I'd actually prefer that 👍 ...that would allow us to add other groups/types of HTML elements in the future if needed. It's much easier to add things to an array parameter than to add/remove/reorder parameters.

**Comment by BWPanda** [comment]
Done! PR updated. I include 'div' in both 'block' and 'layout', and this works nicely as if you specify both block and layout, the duplicates will overwrite each other so you still only get one nice list that includes everything necessary.

I just wasn't sure how to handle [this one](https://github.com/backdrop/backdrop/blob/1.x/core/modules/layout/plugins/styles/layout_style_dynamic.inc#L28-L35):
```php
$tag_list = array(
  'div' => 'DIV',
  'aside' => 'ASIDE',
  'section' => 'SECTION',
  'nav' => 'NAV',
  'p' => 'P',
  '' => t('No wrapper'),
);
```

**Comment by klonos** [comment]
Thanks @BWPanda 🙏 ...one minor request to wrap a sting in `t()` and this should be good 👍

**Comment by BWPanda** [comment]
@klonos Thanks! Updated.

**Comment by klonos** [comment]
Sorry for the late response here...

Code looks good 👍 ...I've also tested this and it seems to work as expected, so RTBC.

I can't add a milestone though, since I was the one to add the milestone label.

**Comment by klonos** [comment]
PS: I found a minor pre-existing bug while testing: #4977

**Comment by quicksketch** [comment]
I am ambivalent about this change. We only have element lists in 5 places, and they use 3 different lists of elements. We're not saving code overall: the PR itself has 92 lines added and 60 lines removed. However, I think the ability to change the list of elements might be helpful, and I like moving the list out of Views config and into a hook.

I have a few suggestions:

- The new `hook_html_elements_alter()` needs to be documented.
- Rather than making `backdrop_html_elements()` take an array (when 4 out of 5 places only use a single group), perhaps make it take a single string. If you need multiple groups you could combine the results.

Instead of:
```
backdrop_html_elements(array('block', 'inline'))
```

You could use:
```
backdrop_html_elements('block') + backdrop_html_elements('inline')
```

The `field_rewrite_elements` config value seems like something that we could remove entirely. The setting has always been hidden as long as it has been in Backdrop. It was added as part of the original Views port, apparently part of combining CTools into core.

So if we proceed, I think it would be cleanest for us to remove the setting, make an update hook, and a changelog.

**Comment by BWPanda** [comment]
I happy for someone else to take this over now.

**Comment by jenlampton** [comment]
> We're not saving code overall: the PR itself has 92 lines added and 60 lines removed.

Maybe not if you look at core alone. But every contrib module that wants to provide a list of HTML elements also needs to duplicate that effort. It would be much easier if there were a Backdrop internal function we could use when we needed to offer them.

> The field_rewrite_elements config value seems like something that we could remove entirely. 

We left it there for Backwards compatibility. If someone had upgraded from Drupal 7 with a customized value stored there, removing it would change the behavior of their site along with this update. Since there are still sites moving from Drupal 7 to Backdrop, I think it's important to maintain backwards compatibility. How about we mark it deprecated and remove in 2.x?

**Comment by quicksketch** [comment]
> If someone had upgraded from Drupal 7 with a customized value stored there, removing it would change the behavior of their site along with this update. Since there are still sites moving from Drupal 7 to Backdrop, I think it's important to maintain backwards compatibility.

This value isn't upgraded from Drupal 7 at all in our current update hooks. It's a hidden variable in D7 as well (no UI), which is why I think that we could probably remove it entirely.

**Comment by jenlampton** [comment]
> This value isn't upgraded from Drupal 7 at all in our current update hooks.

It didn't happen when views settings were converted to config? I think that sounds like another bug, and one that will become increasingly important as more people move from Drupal to Backdrop (and it's more likely that we'll get formerly-D7 sites that have changed it)

> It's a hidden variable in D7 as well (no UI), which is why I think that we could probably remove it entirely.

So you are recommending that people who changed the value in Drupal 7 (likely via a custom update hook, as I did) will need to now implement an alter hook for their Backdrop site to have the same behavior?

I could get behind that as long as the backdrop defaults are significantly better than what was offered in Drupal 7 views.

I am a little bit concerned about adding additional work to the Drupal -> Backdrop upgrade because Backwards compatibility is supposed to be "important". Fixing our update would be less work for core than even a single new hook being added on a that's being updated. I'll create a separate issue for that fix.
